rust/chains/abacus-ethereum/src/multisig_ism.rs

```diff
@@ -0,0 +1,160 @@
+#![allow(clippy::enum_variant_names)]
+#![allow(missing_docs)]
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use abacus_core::accumulator::merkle::Proof;
+use async_trait::async_trait;
+use ethers::abi::Token;
+use ethers::providers::Middleware;
+use ethers::types::{Selector, H160, H256, U256};
+use eyre::Result;
+
+use abacus_core::{
+    AbacusAbi, AbacusContract, ChainCommunicationError, ContractLocator, MultisigIsm,
+    MultisigSignedCheckpoint,
+};
+
+use crate::contracts::multisig_ism::{
+    MultisigIsm as EthereumMultisigIsmInternal, MULTISIGISM_ABI,
+};
+use crate::trait_builder::MakeableWithProvider;
+
+impl<M> std::fmt::Display for EthereumMultisigIsmInternal<M>
+where
+    M: Middleware,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+pub struct MultisigIsmBuilder {}
+
+impl MakeableWithProvider for MultisigIsmBuilder {
+    type Output = Box<dyn MultisigIsm>;
+
+    fn make_with_provider<M: Middleware + 'static>(
+        &self,
+        provider: M,
+        locator: &ContractLocator,
+    ) -> Self::Output {
+        Box::new(EthereumMultisigIsm::new(Arc::new(provider), locator))
+    }
+}
+
+/// A reference to an MultisigIsm contract on some Ethereum chain
+#[derive(Debug)]
+pub struct EthereumMultisigIsm<M>
+where
+    M: Middleware,
+{
+    contract: Arc<EthereumMultisigIsmInternal<M>>,
+    #[allow(dead_code)]
+    domain: u32,
+    chain_name: String,
+    #[allow(dead_code)]
+    provider: Arc<M>,
+}
+
+impl<M> EthereumMultisigIsm<M>
+where
+    M: Middleware + 'static,
+{
+    /// Create a reference to a mailbox at a specific Ethereum address on some
+    /// chain
+    pub fn new(provider: Arc<M>, locator: &ContractLocator) -> Self {
+        Self {
+            contract: Arc::new(EthereumMultisigIsmInternal::new(
+                &locator.address,
+                provider.clone(),
+            )),
+            domain: locator.domain,
+            chain_name: locator.chain_name.to_owned(),
+            provider,
+        }
+    }
+}
+
+impl<M> AbacusContract for EthereumMultisigIsm<M>
+where
+    M: Middleware + 'static,
+{
+    fn chain_name(&self) -> &str {
+        &self.chain_name
+    }
+
+    fn address(&self) -> H256 {
+        self.contract.address().into()
+    }
+}
+
+#[async_trait]
+impl<M> MultisigIsm for EthereumMultisigIsm<M>
+where
+    M: Middleware + 'static,
+{
+    /// Returns the metadata needed by the contract's verify function
+    async fn format_metadata(
+        &self,
+        checkpoint: &MultisigSignedCheckpoint,
+        proof: Proof,
+    ) -> Result<Vec<u8>, ChainCommunicationError> {
+        let threshold = self.threshold(checkpoint.checkpoint.mailbox_domain).await?;
```

Comment:
I think we should try to avoid putting `eth_call`s in this function and instead load the threshold / validators as configs, mostly to reduce the # of RPCs we do.

Atm if the serial submitter's retry queue has messages in it, every second it'll try to evaluate a message for processing. The way it does that atm is by just estimating gas for the `Mailbox.process` call, meaning it'll need to call `format_metadata`. Adding these 2 eth_calls that'll be done each second for each destination mailbox with a retry queue could increase the RPCs we do by a decent amount. Curious how you feel about making it config based and just needing to change the config when validator sets change. We need to update the config anyways to add validator S3 buckets

Reply:
I think eventually* we'll want the following:
1. A process in the relayer that indexes `ValidatorRegistered` events and populates a mapping of validator address -> signed checkpoint location
2. When relaying a message, the relayer queries the recipient for their ISM, and then the ISM for their threshold/validator set
3. The relayer uses (1) and (2) to format the metadata needed to process the message

You can reduce the RPC load for messages stuck in the retry queue by caching the responses to (2) and only re-querying every so often.

But you're right that probably it doesn't make sense to do the eth_calls now more than once given that we don't functionally support sovereign consensus in the agents yet

* Eventually we'll of course want even more than this when the relayer supports multiple ISM types, what I described is a vision for MultisigIsm only.

Reply:
Dumb rust question, what's the best way to do this?

I tried adding `threshold` and `validators` hashmaps to the structs and then populating them if there wasn't an entry to `domain` but I can't seem to do that because `EthereumMultisigIsm` is not mutable.

Reply:
mm I think there are a couple ways
One could be something like what I think you're saying, where you have a cache internal to `EthereumMultisigIsm` that fetches the threshold and validators from the smart contract if it's not already known. You could maybe use a OnceCell for this, or if that doesn't work a Mutex
But I'm not certain this is the best way because of how we are likely to want to deal with making changes to the validator set. I think before when we wanted to increase the quorum threshold from e.g. 2 to 3, we first changed the relayers to pretend like the threshold was 3, and then made the on-chain change to move from 2 to 3 so that we never had downtime. So maybe relying upon configuration will help us handle these changes more easily?

If you want to go with the configuration route, passing in the config values initially when creating the `MultisigIsmBuilder` might be the move

Reply:
Hmm yeah that's a good point..

That said I'd like to move towards relying less on hard coding config parameters that can change dynamically, not more, since to support sovereign consensus we'll want the relayer to not require any configuration of the validator set(s).

I'm going to try having a cache that gets refreshed if:
- Gas estimation for message processing fails
- The last time we refreshed the cache was > X seconds ago

Reply:
@mattiecnvr it looks like maybe I want the following?

```#[memoize(SharedCache, TimeToLive: Duration::from_secs(60))]```

Reply:
Hmm maybe that doesn't work, getting:
```
error: Cannot memoize method (self-receiver) without arguments!
```

Reply:
> Gas estimation for message processing fails

This part wound up being a bit more work than anticipated b/c in the gelato submitter we don't actually estimate gas. I punted for now since all of this is somewhat temporary anyways until we have more complete sovereign consensus support.

Reply:
nice, sgtm